lib/TPP/Transforms/VectorContractToOuterproduct.cpp

```diff
@@ -0,0 +1,199 @@
+
+    auto maps = contractOp.getIndexingMapsArray();
+    if (llvm::any_of(
+            maps, [](AffineMap map) { return !map.isProjectedPermutation(); }))
+      return rewriter.notifyMatchFailure(contractOp, "Unexpected map");
+
+    // Check if this is a matrix multiply
+    auto iteratorTypes = contractOp.getIteratorTypesArray();
+    if (iteratorTypes.size() < 3 ||
+        iteratorTypes[iteratorTypes.size() - 3] !=
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 2] !=
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 1] !=
+            vector::IteratorType::reduction)
+      return rewriter.notifyMatchFailure(contractOp, "Not a gemm");
+
+    Location loc = contractOp.getLoc();
+    Value lhs = contractOp.getLhs();
+    Value rhs = contractOp.getRhs();
+    Value acc = contractOp.getAcc();
+
+    // Find the original tensor operands
+    auto lhsDefiningOp = lhs.getDefiningOp<vector::TransferReadOp>();
+    auto rhsDefiningOp = rhs.getDefiningOp<vector::TransferReadOp>();
+    if (!lhsDefiningOp || !rhsDefiningOp)
+      return failure();
+
+    Value lhsTensor = lhsDefiningOp.getSource();
+    Value rhsTensor = rhsDefiningOp.getSource();
+
+    auto lhsType = cast<ShapedType>(lhsTensor.getType());
+    auto rhsType = cast<ShapedType>(rhsTensor.getType());
+    auto accType = cast<VectorType>(acc.getType());
+
+    // Handle 3D subviews
+    auto mapLHS = maps[0];
+    unsigned matrixDimsLHS = mapLHS.getNumResults();
+
+    int64_t M = accType.getDimSize(0);
+    int64_t N = accType.getDimSize(1);
+    int64_t K = !isTransposed(mapLHS, matrixDimsLHS)
+                    ? lhsType.getDimSize(lhsType.getRank() - 1)
+                    : lhsType.getDimSize(lhsType.getRank() - 2);
+
+    // Create constants
+    Value c0 = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value c1 = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value cK = rewriter.create<arith::ConstantIndexOp>(loc, K);
+
+    auto elementType = lhsType.getElementType();
+    FloatType floatType = cast<FloatType>(elementType);
+    Value f0 = rewriter.create<arith::ConstantFloatOp>(
+        loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+
+    // Create the outer scf.for loop
+    auto forOp = rewriter.create<scf::ForOp>(
+        loc, c0, cK, c1, ValueRange{acc},
+        [&](OpBuilder &nestedBuilder, Location nestedLoc, Value iv,
+            ValueRange iterArgs) {
+          // Prepare indices for 3D subviews
+          SmallVector<Value, 3> lhsIndices, rhsIndices;
+          for (int i = 0; i < lhsType.getRank() - 2; ++i)
+            lhsIndices.push_back(c0);
+          lhsIndices.push_back(c0);
+          lhsIndices.push_back(iv);
+
+          for (int i = 0; i < rhsType.getRank() - 2; ++i)
+            rhsIndices.push_back(c0);
+          rhsIndices.push_back(iv);
+          rhsIndices.push_back(c0);
```

Comment:
Does it account for shifting K-dim position in case the operands are transposed?

Reply:
No, but it seems I need to reorder the indices accordingly. Thanks.